core/arch/arm/plat-stm32mp1/plat_huk.c

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, Linaro Limited
+ * Copyright (c) 2022, Foundries.io Limited
+ */
+
+#include <assert.h>
+#include <config.h>
+#include <crypto/crypto.h>
+#include <drivers/stm32_bsec.h>
+#include <kernel/tee_common_otp.h>
+#include <platform_config.h>
+#include <string.h>
+
+#if !defined(CFG_STM32MP1_HUK_BSEC_BASE)
+#error "CFG_STM32MP1_HUK_BSEC_BASE undefined"
+#endif
+
+static struct {
+	uint8_t key[HW_UNIQUE_KEY_LENGTH];
+	bool ready;
+} huk;
+
+TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
+{
+	const uint32_t bsec_base = CFG_STM32MP1_HUK_BSEC_BASE;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t key[16] = { 0 };
+	uint8_t tag[16] = { 0 };
+	uint8_t uid[16] = { 0 };
+	uint8_t nce[12] = {
+		0xd2, 0x45, 0x0e, 0x07, 0xea, 0x5d,
+		0xe0, 0x42, 0x6c, 0x0f, 0xa1, 0x33,
+	};
+	size_t enc_len = sizeof(huk.key);
+	size_t tag_len = sizeof(tag);
+	size_t uid_len = sizeof(uid);
+	uint32_t word = 0;
+	void *ctx = NULL;
+	size_t n = 0;
+
+	if (huk.ready)
+		goto out;
+
+	for (n = 0; n < sizeof(key) / sizeof(uint32_t); n++) {
+		res = stm32_bsec_shadow_register(bsec_base + n);
+		if (res)
+			return res;
+
+		if (IS_ENABLED(CFG_STM32MP1_HUK_BSEC_SHADOW_TESTKEY)) {
+			res = stm32_bsec_write_otp(0, bsec_base + n);
+			if (res)
+				return res;
+		}
+
+		res = stm32_bsec_read_otp(&word, bsec_base + n);
+		if (res)
+			return res;
+
+		memcpy(key + n * sizeof(uint32_t), &word, sizeof(uint32_t));
+	}
+
+	for (n = 0; n < (UID2_OTP - UID0_OTP + 1); n++) {
```

Comment:
I just shrinked the word nonce which could be any value and doesn't need to be secret. Maybe I should leave it as nonce?

Reply:
Right, AMD/Xilinx proposed this scheme to protect public uniqueness (ie SoC ID) via a secret key- it is what is being used in Zynqmp and in Versal ACAP so I thikn it makes sense to use it globally.

It makes provisioning after manufacturing easier since the secret key can be shared across all devices.

if we used SHA and the SoC Id became known then we would have an issue (I thikn this can be read from U-boot?) ; but then again derivation is disabled by default so for the simple scheme the user can read the uniqueness from those OTPs instead

Reply:
Ok. Having this derivation specifically under `CFG_STM32MP15_HUK_BSEC_DERIVE_DEV_UID` looks right to me.